### PR TITLE
broadcast transaction conversions refactor

### PR DESCRIPTION
**Pull request by dbejarano820**
Dev work to refactor the Broadcasted Transaction conversions to Transactions into try_from implementations on the primitive types themselves in order to be more rustlike.

# Pull Request type
- Refactoring (no functional changes, no API changes)

## What is the current behavior?
Uses functions from utils to convert Broadcast Transactions into normal transactions

Resolves: #552 

## What is the new behavior?
Removes the functions
- to_declare_tx
- to_deploy_account_tx
- to_invoke_tx

Try_from implementations added

## Does this introduce a breaking change?
No

### CODE REVIEW

**Comment by LucasLvy** [comment]
fix format and clippy plz

**Comment by dbejarano820** [comment]
> fix format and clippy plz

done!

File: crates/primitives/starknet/src/transaction/types.rs [hunk 1]
```diff
@@ -212,6 +221,55 @@ impl DeclareTransaction {
     }
 }
 
+#[cfg(feature = "std")]
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/types.rs:224]: keep this no std compatible

File: crates/primitives/starknet/src/transaction/types.rs [hunk 2]
```diff
@@ -212,6 +221,55 @@ impl DeclareTransaction {
...
+impl TryFrom<BroadcastedDeclareTransaction> for DeclareTransaction {
+    type Error = anyhow::Error;
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/types.rs:226]: don't use anyhow as error

File: crates/primitives/starknet/src/transaction/types.rs [hunk 3]
```diff
@@ -383,6 +441,31 @@ impl InvokeTransaction {
     }
 }
 
+#[cfg(feature = "std")]
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/types.rs:444]: keep it no std

File: crates/primitives/starknet/src/transaction/types.rs [hunk 4]
```diff
@@ -383,6 +441,31 @@ impl InvokeTransaction {
...
+impl TryFrom<BroadcastedInvokeTransaction> for InvokeTransaction {
+    type Error = anyhow::Error;
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/types.rs:446]: don't use anyhow

File: crates/primitives/starknet/src/transaction/types.rs [hunk 5]
```diff
@@ -435,6 +518,45 @@ impl TryFrom<Transaction> for DeployAccountTransaction {
     }
 }
 
+#[cfg(feature = "std")]
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/types.rs:521]: keep it no std

File: crates/primitives/starknet/src/transaction/types.rs [hunk 6]
```diff
@@ -435,6 +518,45 @@ impl TryFrom<Transaction> for DeployAccountTransaction {
...
+impl TryFrom<BroadcastedDeployAccountTransaction> for DeployAccountTransaction {
+    type Error = anyhow::Error;
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/types.rs:523]: don't use anyhow

File: crates/primitives/starknet/src/transaction/utils.rs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+use alloc::collections::BTreeMap;
+use alloc::vec::Vec;
+
+use frame_support::BoundedVec;
+#[cfg(feature = "std")]
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/utils.rs:5]: make it no std

File: crates/primitives/starknet/src/transaction/utils.rs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+use starknet_core::types::{LegacyContractEntryPoint, LegacyEntryPointsByType};
+
+use crate::execution::types::{EntryPointTypeWrapper, EntryPointWrapper, MaxEntryPoints};
+
+/// Returns a btree map of entry point types to entrypoint from deprecated entry point by type
+#[cfg(feature = "std")]
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/utils.rs:11]: why did you move that ?

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: crates/primitives/starknet/src/transaction/types.rs [hunk 9]
```diff
@@ -212,6 +221,55 @@ impl DeclareTransaction {
...
+                    max_fee: Felt252Wrapper::from(declare_tx_v1.max_fee),
+                    signature,
+                    contract_class: ContractClassWrapper {
+                        program: program
+                            .try_into()
+                            .map_err(|_| anyhow!("Failed to convert program to program wrapper"))?,
+                        entry_points_by_type: BoundedBTreeMap::try_from(to_btree_map_entrypoints(
+                            declare_tx_v1.contract_class.entry_points_by_type.clone(),
+                        ))
+                        .unwrap(),
```
[the remark this answers is not in the document]
Comment (dbejarano820) [reply crates/primitives/starknet/src/transaction/types.rs:263]: unwrap is part of the existing implementation for this, I can remove it

File: crates/primitives/starknet/src/transaction/utils.rs (hunk 8 above)
Comment (dbejarano820) [reply crates/primitives/starknet/src/transaction/utils.rs:11]: because it's used in `to_declare_tx` and since we are moving that logic to a `try_from` implementation in a primitive, I thing to bring that function over as well. Correct me if I'm wrong

File: crates/primitives/starknet/src/transaction/types.rs (hunk 6 above)
Comment (dbejarano820) [reply crates/primitives/starknet/src/transaction/types.rs:523]: The current implementation uses anyhow, should I make a new Error enum or use the existing `TransactionExecutionErrorWrapper` in types.rs

File: crates/primitives/starknet/src/transaction/types.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (LucasLvy) [reply crates/primitives/starknet/src/transaction/types.rs:263]: yep remove it plz

File: crates/primitives/starknet/src/transaction/utils.rs (hunk 7 above)
Comment (LucasLvy) [reply crates/primitives/starknet/src/transaction/utils.rs:5]: it should yes

File: crates/primitives/starknet/src/transaction/utils.rs (hunk 8 above)
Comment (LucasLvy) [reply crates/primitives/starknet/src/transaction/utils.rs:11]: makes sense. But make sure that you rebased, this function doesn't exist anymore, it has been changed to use hashmaps

File: crates/primitives/starknet/src/transaction/types.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (dbejarano820) [reply crates/primitives/starknet/src/transaction/types.rs]: yup, wasn't aware that function had been changed altogether

File: crates/primitives/starknet/src/transaction/utils.rs (hunk 7 above)
Comment (glihm) [reply crates/primitives/starknet/src/transaction/utils.rs:2]: @LucasLvy it is worth double check with Jonathan, but the last time I asked him, `starknet-core` is not yet `no-std` compatible. :/
And all the types in use here come from `starknet-core`.

Comment (dbejarano820) [reply crates/primitives/starknet/src/transaction/utils.rs:2]: Once it's `no-std` compatible we can make the relevant changes here

**Review by LucasLvy — APPROVED** [review]
